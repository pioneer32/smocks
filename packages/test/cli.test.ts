import fetch, { RequestInit } from 'node-fetch';
import path from 'node:path';
import { ChildProcessByStdio, spawn } from 'node:child_process';

import https from 'node:https';

const compiledCliScriptPath = path.resolve(__dirname, '../smocks/dist/cli.cjs');

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

const request = async (url: string, init?: RequestInit) => {
  // @ts-ignore
  const res = await fetch(url, init);

  return {
    status: res.status,
    headers: [...res.headers.entries()].filter(([name]) => name !== 'date' && name !== 'etag').map(([name, value]) => `${name}: ${value}`),
    text: await res.text(),
  };
};

describe('CLI', () => {
  describe('static config', () => {
    const configFile = path.resolve(__dirname, 'config.static.cjs');
    let childProcess: ChildProcessByStdio<null, null, null>;
    beforeAll(async () => {
      console.log('Starting Smocks CLI as a child process...')
      childProcess = spawn('node', [compiledCliScriptPath, 'start', '-c', configFile], {
        stdio: ['inherit', 'inherit', 'inherit'],
      });
      await sleep(2000);
    });
    afterAll(async () => {
      childProcess.kill('SIGHUP');
    });

    it('works', async () => {
      expect(await await request('http://localhost:3001/session/default')).toMatchSnapshot('session-details');
      expect(await request('http://localhost:3000/login')).toMatchSnapshot('response');
    });
  });

  describe('dynamic config', () => {
    const configFile = path.resolve(__dirname, 'config.dynamic.cjs');
    const agent = new https.Agent({ rejectUnauthorized: false });
    let childProcess: ChildProcessByStdio<null, null, null>;
    beforeAll(async () => {
      childProcess = spawn('node', [compiledCliScriptPath, 'start', '-c', configFile], {
        stdio: ['inherit', 'inherit', 'inherit'],
      });
      await sleep(2000);
    });
    afterAll(async () => {
      childProcess.kill('SIGHUP');
    });

    it('works', async () => {
      expect(await await request('https://localhost:3001/session/default', { agent })).toMatchSnapshot('session-details');
      expect(await request('https://localhost:3000/login', { agent })).toMatchSnapshot('response');
    });
  });
});
