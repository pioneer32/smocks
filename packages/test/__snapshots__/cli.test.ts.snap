// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI dynamic config works: response 1`] = `
{
  "headers": [
    "access-control-allow-headers: *",
    "access-control-allow-methods: OPTIONS,POST,GET",
    "access-control-allow-origin: *",
    "connection: close",
    "content-length: 239",
    "content-type: text/html; charset=utf-8",
    "x-powered-by: Express",
  ],
  "status": 200,
  "text": "<!DOCTYPE html>
<html>
    <head>
        <title>Pseudo IAM login page</title>
    </head>
    <body>
        <h1>Welcome to Pseudo login page</h1>
        <a href="/?state=fooBar&code=cb485bf">click here to log in</a>
    </body>
</html>
",
}
`;

exports[`CLI dynamic config works: session-details 1`] = `
{
  "headers": [
    "connection: close",
    "content-length: 15",
    "content-type: application/json; charset=utf-8",
    "x-powered-by: Express",
  ],
  "status": 200,
  "text": "{"requests":[]}",
}
`;

exports[`CLI static config works: response 1`] = `
{
  "headers": [
    "access-control-allow-headers: *",
    "access-control-allow-methods: OPTIONS,POST,GET",
    "access-control-allow-origin: *",
    "connection: close",
    "content-length: 239",
    "content-type: text/html; charset=utf-8",
    "x-powered-by: Express",
  ],
  "status": 200,
  "text": "<!DOCTYPE html>
<html>
    <head>
        <title>Pseudo IAM login page</title>
    </head>
    <body>
        <h1>Welcome to Pseudo login page</h1>
        <a href="/?state=fooBar&code=cb485bf">click here to log in</a>
    </body>
</html>
",
}
`;

exports[`CLI static config works: session-details 1`] = `
{
  "headers": [
    "connection: close",
    "content-length: 15",
    "content-type: application/json; charset=utf-8",
    "x-powered-by: Express",
  ],
  "status": 200,
  "text": "{"requests":[]}",
}
`;
